design/notifications.md

```diff
@@ -0,0 +1,405 @@
+for this service despite the increased development time. If we end up having more than a few deployments in
+resource-constrained environments then the reduction in deployment headaches will more than make up for the development
+time.
+
+## Component Detail
+
+### Event Recorder
+
+The event recorder is the first link in the notification processing chain. It listens to the AMQP exchange for incoming
+messages indicating that an event has occurred. For the time being, this routing key will follow the format
+`events.{category}.update.{update-type}`. The category component of the routing key will describe the category of the
+event. The update type refers to the type of update that occurred. For example, if an app was published then the
+category would be `apps` and the update type would be `published`, meaning that the routing key would be
+`events.apps.update.published`. Having different routing keys for different types of updates will eventually allow us to
+easily support different handlers for different types of updates if we wish to do so at some point in the
+future. Backwards compatibilty will be maintained by using `notification` as the category and placing the current
+notification type in the update type field, so all app updates that are posted to the notification publishing route in
+the REST API will automatically have a routing key of `events.notification.update.apps`. This may be slightly confusing,
+but it's a good way to ensure that all backwards compatible notifications are managed by the same handler.
+
+Once the event recorder receives a message, it has to decide what if anything to do with the message. If the event
+recorder decides that no user has to be informed of the event then it will simply acknowledge the message and wait for
+the next one to arrive. If it decides that one or more users have to be informed of the event then processing will
+continue. This decision will be influenced by the category and the update type, so each handler will likely make this
+decision in a unique manner except. In the case of the backwards compatible mode, the decision will be made in the same
+way as in the curernt notification agent out of necessity.
+
+The next step is to determine who should receive the notifications. This is another decision that is likely to vary
+depending on category and update type. The handler for backwards compatible notifications will only send the
+notification to the user mentioned in the `user` field of the notification request.
+
+Next, the event recorder will have to determine whether or not an email should be sent. And once again, this decision
+will probably vary depending on the category and update type. Eventually, we'll probably also want to have a
+subscription mechanism so that users can choose which types of notifications they get emails for. For the backwards
+compatibility case, however, the user will only receive an email if the notificaton request indicates that an email
+should be sent.
```

Comment:
Another thing we may want to consider eventually, which I think this design enables, is email batching, especially in cases where many similar notifications come in at once.